File: Ui/Scene2D.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Numerics;
using Silk.NET.Vulkan;
using SETUE.Core;
using SETUE.ECS;
using SETUE.RenderEngine;
using SETUE.Systems;
using static SETUE.Vulkan;
using VkBuffer = Silk.NET.Vulkan.Buffer;

namespace SETUE.Scene
{
    public class DrawCommand2D
    {
        public Matrix4x4 Transform;
        public Vector4 Color;
        public string PipelineId = "";
        public VkBuffer VertexBuffer;
        public VkBuffer IndexBuffer;
        public uint IndexCount;
        public int Layer;
        public bool IsText;
        public string FontId = "";
        public bool UseScissor;
        public Rect2D Scissor;
    }

    public class Scene2DRule
    {
        public string Id = "";
        public bool Enabled;
        public int Order;
        public string DataSource = "";
        public string ItemFilter = "";
        public string MeshSource = "";
        public string TransformSource = "";
        public string ColorSource = "";
        public bool UseScissor;
    }

    public static class Scene2D
    {
        private static List<Scene2DRule> _rules = new();
        public static List<DrawCommand2D> Commands { get; private set; } = new();
        private static Dictionary<int, Rect2D> _panelClipRects = new();

        public static void Load()
        {
            string path = "Ui/Scene2D.csv";
            if (!File.Exists(path))
            {
                Console.WriteLine($"[Scene2D] Missing {path}");
                return;
            }

            var lines = File.ReadAllLines(path);
            var headers = lines[0].Split(',');

            int idxId = Array.IndexOf(headers, "id");
            int idxEnabled = Array.IndexOf(headers, "enabled");
            int idxOrder = Array.IndexOf(headers, "order");
            int idxDataSource = Array.IndexOf(headers, "data_source");
            int idxFilter = Array.IndexOf(headers, "item_filter");
            int idxMesh = Array.IndexOf(headers, "mesh_source");
            int idxTransform = Array.IndexOf(headers, "transform_source");
            int idxColor = Array.IndexOf(headers, "color_source");
            int idxUseScissor = Array.IndexOf(headers, "use_scissor");

            _rules.Clear();
            for (int i = 1; i < lines.Length; i++)
            {
                var line = lines[i].Trim();
                if (string.IsNullOrEmpty(line) || line.StartsWith('#')) continue;
                var p = line.Split(',');
                string Get(int idx) => idx >= 0 && idx < p.Length ? p[idx].Trim() : "";

                _rules.Add(new Scene2DRule
                {
                    Id = Get(idxId),
                    Enabled = Get(idxEnabled).ToLower() == "true",
                    Order = int.TryParse(Get(idxOrder), out var o) ? o : 0,
                    DataSource = Get(idxDataSource),
                    ItemFilter = Get(idxFilter),
                    MeshSource = Get(idxMesh),
                    TransformSource = Get(idxTransform),
                    ColorSource = Get(idxColor),
                    UseScissor = Get(idxUseScissor).ToLower() == "true"
                });
            }

            _rules.Sort((a, b) => a.Order.CompareTo(b.Order));
            Console.WriteLine($"[Scene2D] Loaded {_rules.Count} rules");
        }

        public static void Update()
        {
            float sw = SwapExtent.Width;
            float sh = SwapExtent.Height;
            Commands.Clear();
            _panelClipRects.Clear();

            var world = Object.ECSWorld;

            ComputePanelClipRects(world, sw, sh);

            foreach (var rule in _rules)
            {
                if (!rule.Enabled) continue;

                switch (rule.DataSource)
                {
                    case "panels":
                        foreach (var (e, transform, panel) in world.Query<TransformComponent, PanelComponent>())
                        {
                            if (!panel.Visible) continue;
                            var material = world.GetComponent<MaterialComponent>(e);

                            float left   = transform.Position.X - transform.Scale.X * 0.5f;
                            float top    = transform.Position.Y - transform.Scale.Y * 0.5f;
                            float width  = transform.Scale.X;
                            float height = transform.Scale.Y;

                            float x = (left / sw) * 2f - 1f;
                            float y = 1f - (top / sh) * 2f;
                            float w = (width / sw) * 2f;
                            float h = (height / sh) * 2f;

                            float cx = x + w * 0.5f;
                            float cy = y - h * 0.5f;

                            Matrix4x4 panelTransform =
                                Matrix4x4.CreateScale(w, -h, 1f) *
                                Matrix4x4.CreateTranslation(cx, cy, 0f);

                            string meshId = rule.MeshSource.StartsWith("fixed:") ? rule.MeshSource[6..] : rule.MeshSource;
                            if (MeshBuffer.Get(meshId, out var vbuf, out var ibuf, out uint idxCount))
                            {
                                Rect2D? clipRect = GetClipRectForPanel(world, e, panel.Id);
                                Commands.Add(new DrawCommand2D
                                {
                                    Transform = panelTransform,
                                    Color = material.Color,
                                    PipelineId = StringRegistry.GetString(material.PipelineId),
                                    VertexBuffer = vbuf,
                                    IndexBuffer = ibuf,
                                    IndexCount = idxCount,
                                    Layer = panel.Layer,
                                    IsText = false,
                                    UseScissor = rule.UseScissor && clipRect.HasValue,
                                    Scissor = clipRect ?? default
                                });
                            }
                            else
                            {
                                Console.WriteLine($"[Scene2D] Mesh '{meshId}' not found for panel '{StringRegistry.GetString(panel.Id)}'");
                            }
                        }
                        break;

                    case "texts":
                        foreach (var (e, text, transform) in world.Query<TextComponent, TransformComponent>())
                        {
                            if (text.PanelId != 0)
                            {
                                bool panelVisible = false;
                                world.ForEach<PanelComponent>((Entity pe) =>
                                {
                                    var p = world.GetComponent<PanelComponent>(pe);
                                    if (p.Id == text.PanelId && p.Visible)
                                        panelVisible = true;
                                });
                                if (!panelVisible) continue;
                            }

                            string fontIdStr = StringRegistry.GetString(text.FontId);
                            if (string.IsNullOrEmpty(fontIdStr)) fontIdStr = "default";
                            var font = SETUE.UI.Fonts.Get(fontIdStr);
                            if (font == null) continue;

                            BuildTextBuffers(text, transform, font, sw, sh, out var vbuf, out var ibuf, out uint idxCount);
                            if (idxCount > 0)
                            {
                                // For text, use the panel ID to find clip rect (text doesn't have DragComponent)
                                Rect2D? clipRect = text.PanelId != 0 ? GetClipRectForPanelId(world, text.PanelId) : null;
                                Commands.Add(new DrawCommand2D
                                {
                                    Transform = Matrix4x4.Identity,
                                    Color = text.Color,
                                    PipelineId = "text_pipeline",
                                    VertexBuffer = vbuf,
                                    IndexBuffer = ibuf,
                                    IndexCount = idxCount,
                                    Layer = text.Layer,
                                    IsText = true,
                                    FontId = fontIdStr,
                                    UseScissor = rule.UseScissor && clipRect.HasValue,
                                    Scissor = clipRect ?? default
                                });
                            }
                        }
                        break;
                }
            }

            Commands.Sort((a, b) => a.Layer.CompareTo(b.Layer));
        }

        private static void ComputePanelClipRects(World world, float screenWidth, float screenHeight)
        {
            // Identify containers: panels with ClipChildren == true
            HashSet<int> containerIds = new();
            world.ForEach<PanelComponent>((Entity e) =>
            {
                var panel = world.GetComponent<PanelComponent>(e);
                if (panel.ClipChildren)
                    containerIds.Add(panel.Id);
            });

            foreach (var containerId in containerIds)
            {
                world.ForEach<PanelComponent>((Entity e) =>
                {
                    var panel = world.GetComponent<PanelComponent>(e);
                    if (panel.Id == containerId && panel.Visible)
                    {
                        var trans = world.GetComponent<TransformComponent>(e);
                        float left   = trans.Position.X - trans.Scale.X * 0.5f;
                        float top    = trans.Position.Y - trans.Scale.Y * 0.5f;
                        float width  = trans.Scale.X;
                        float height = trans.Scale.Y;

                        int scissorX = (int)Math.Max(0, left);
                        int scissorY = (int)Math.Max(0, top);
                        int scissorW = (int)Math.Min(width, screenWidth - scissorX);
                        int scissorH = (int)Math.Min(height, screenHeight - scissorY);

                        if (scissorW > 0 && scissorH > 0)
                        {
                            _panelClipRects[containerId] = new Rect2D(
                                new Offset2D(scissorX, scissorY),
                                new Extent2D((uint)scissorW, (uint)scissorH));
                        }
                    }
                });
            }
        }

        /// <summary>
        /// Returns a scissor rectangle for the given panel entity if it should be clipped.
        /// A panel is clipped ONLY if it has a DragComponent that references a parent,
        /// AND that parent has ClipChildren == true. The clip rect of the parent is used.
        /// </summary>
        private static Rect2D? GetClipRectForPanel(World world, Entity entity, int panelId)
        {
            // Does this entity have a DragComponent with a valid ParentNameId?
            if (!world.HasComponent<DragComponent>(entity))
                return null;

            var drag = world.GetComponent<DragComponent>(entity);
            int parentId = drag.ParentNameId;
            if (parentId == 0)
                return null;

            // If the parent is a container (ClipChildren == true), return its clip rect
            if (_panelClipRects.TryGetValue(parentId, out var rect))
                return rect;

            return null;
        }

        /// <summary>
        /// For text entities (which lack DragComponent), we use the associated PanelId
        /// to find the panel's clip rect. The text inherits clipping from its panel.
        /// </summary>
        private static Rect2D? GetClipRectForPanelId(World world, int panelId)
        {
            // If the panel itself is a container, use its clip rect
            if (_panelClipRects.TryGetValue(panelId, out var rect))
                return rect;

            // Otherwise, we need to find the panel entity and check its DragComponent parent
            Entity? panelEntity = null;
            world.ForEach<PanelComponent>((Entity e) =>
            {
                var p = world.GetComponent<PanelComponent>(e);
                if (p.Id == panelId)
                    panelEntity = e;
            });

            if (panelEntity.HasValue)
            {
                return GetClipRectForPanel(world, panelEntity.Value, panelId);
            }

            return null;
        }

        private static void BuildTextBuffers(TextComponent text, TransformComponent transform, SETUE.UI.Font font, float sw, float sh,
            out VkBuffer vbuf, out VkBuffer ibuf, out uint indexCount)
        {
            vbuf = default; ibuf = default; indexCount = 0;
            string content = StringRegistry.GetString(text.ContentId);
            if (string.IsNullOrEmpty(content)) return;

            List<float> verts = new();
            List<uint> indices = new();
            uint idxOffset = 0;

            float startX = transform.Position.X;
            float startY = transform.Position.Y;

            float totalWidth = 0f;
            foreach (char c in content)
                if (font.Glyphs.TryGetValue(c, out var g))
                    totalWidth += g.AdvanceX;

            float ascent = font.Ascent;
            float descent = 0f;
            var descentProp = font.GetType().GetProperty("Descent");
            if (descentProp != null)
                descent = (float)descentProp.GetValue(font);
            else
                descent = ascent * 0.25f;

            float visualHeight = ascent + descent;
            float visualTop = -ascent;
            float visualBottom = descent;

            float centerOffsetX = 0f;
            float centerOffsetY = (visualTop + visualBottom) * 0.5f;

            string alignStr = StringRegistry.GetString(text.Align);
            if (alignStr == "left")
                centerOffsetX = totalWidth * 0.5f;
            else if (alignStr == "center")
                centerOffsetX = 0f;
            else if (alignStr == "right")
                centerOffsetX = -totalWidth * 0.5f;

            float pivotX = startX + centerOffsetX;
            float pivotY = startY + centerOffsetY;

            float baseX = pivotX - totalWidth * 0.5f;
            float baseY = pivotY - centerOffsetY;

            float penX = baseX;

            float cosR = MathF.Cos(text.Rotation * MathF.PI / 180f);
            float sinR = MathF.Sin(text.Rotation * MathF.PI / 180f);

            foreach (char c in content)
            {
                if (!font.Glyphs.TryGetValue(c, out var glyph))
                {
                    penX += 8f;
                    continue;
                }

                float x0 = penX;
                float y0 = baseY - ascent;
                float x1 = x0 + glyph.Width;
                float y1 = y0 + glyph.Height;

                float rx0, ry0, rx1, ry1, rx2, ry2, rx3, ry3;

                if (text.Rotation == 0)
                {
                    rx0 = x0; ry0 = y0;
                    rx1 = x1; ry1 = y0;
                    rx2 = x1; ry2 = y1;
                    rx3 = x0; ry3 = y1;
                }
                else
                {
                    (rx0, ry0) = RotatePoint(x0, y0, pivotX, pivotY, cosR, sinR);
                    (rx1, ry1) = RotatePoint(x1, y0, pivotX, pivotY, cosR, sinR);
                    (rx2, ry2) = RotatePoint(x1, y1, pivotX, pivotY, cosR, sinR);
                    (rx3, ry3) = RotatePoint(x0, y1, pivotX, pivotY, cosR, sinR);
                }

                float nx0 = (rx0 / sw) * 2f - 1f;
                float ny0 = 1f - (ry0 / sh) * 2f;
                float nx1 = (rx1 / sw) * 2f - 1f;
                float ny1 = 1f - (ry1 / sh) * 2f;
                float nx2 = (rx2 / sw) * 2f - 1f;
                float ny2 = 1f - (ry2 / sh) * 2f;
                float nx3 = (rx3 / sw) * 2f - 1f;
                float ny3 = 1f - (ry3 / sh) * 2f;

                verts.AddRange(new[] {
                    nx0, ny0, 0f, glyph.U0, glyph.V0, 0f, 0f, 0f,
                    nx1, ny1, 0f, glyph.U1, glyph.V0, 0f, 0f, 0f,
                    nx2, ny2, 0f, glyph.U1, glyph.V1, 0f, 0f, 0f,
                    nx0, ny0, 0f, glyph.U0, glyph.V0, 0f, 0f, 0f,
                    nx2, ny2, 0f, glyph.U1, glyph.V1, 0f, 0f, 0f,
                    nx3, ny3, 0f, glyph.U0, glyph.V1, 0f, 0f, 0f
                });

                indices.AddRange(new[] {
                    idxOffset, idxOffset+1, idxOffset+2,
                    idxOffset+3, idxOffset+4, idxOffset+5
                });
                idxOffset += 6;

                penX += glyph.AdvanceX;
            }

            if (verts.Count == 0) return;

            ulong vSize = (ulong)(verts.Count * sizeof(float));
            ulong iSize = (ulong)(indices.Count * sizeof(uint));
            Vulkan.CreateBuffer(vSize, BufferUsageFlags.VertexBufferBit,
                MemoryPropertyFlags.HostVisibleBit | MemoryPropertyFlags.HostCoherentBit,
                out vbuf, out var vmem);
            Vulkan.CreateBuffer(iSize, BufferUsageFlags.IndexBufferBit,
                MemoryPropertyFlags.HostVisibleBit | MemoryPropertyFlags.HostCoherentBit,
                out ibuf, out var imem);

            unsafe
            {
                void* mapped;
                Vulkan.VK.MapMemory(Vulkan.Device, vmem, 0, vSize, 0, &mapped);
                fixed (float* src = verts.ToArray())
                    System.Buffer.MemoryCopy(src, mapped, (long)vSize, (long)vSize);
                Vulkan.VK.UnmapMemory(Vulkan.Device, vmem);

                Vulkan.VK.MapMemory(Vulkan.Device, imem, 0, iSize, 0, &mapped);
                fixed (uint* src = indices.ToArray())
                    System.Buffer.MemoryCopy(src, mapped, (long)iSize, (long)iSize);
                Vulkan.VK.UnmapMemory(Vulkan.Device, imem);
            }

            indexCount = (uint)indices.Count;
            Vulkan.DeferFree(vbuf, vmem);
            Vulkan.DeferFree(ibuf, imem);
        }

        private static (float, float) RotatePoint(float x, float y, float cx, float cy, float cosR, float sinR)
        {
            float dx = x - cx;
            float dy = y - cy;
            return (cx + dx * cosR - dy * sinR, cy + dx * sinR + dy * cosR);
        }
    }
}

File: Ui/Scene2D.ts
// Ui/Scene2D.ts – draws panels (tags + static styles), text (anchor‑aware), clip‑child masks
import * as PIXI from 'pixi.js';
import { registerSystemMethod } from '../Core/Scheduler.js';
import { getWorld } from '../Core/GlobalWorld.js';
import { styleArray, layerBuckets } from './Panel.js';
import { Window } from '../Window.js';
import {
  TransformComponent,
  PanelComponent,
  TextComponent,
  Entity,
} from '../Core/ECS.js';

let stage: PIXI.Container | null = null;
const panelSprites = new Map<number, PIXI.Container>();
const textSprites   = new Map<number, PIXI.Text>();
const lastScaleX    = new Map<number, number>();
const lastScaleY    = new Map<number, number>();
const lastColorHex  = new Map<number, number>();

function toHexColor(r: number, g: number, b: number): number {
  return ((Math.round(r * 255) & 0xff) << 16) |
         ((Math.round(g * 255) & 0xff) << 8)  |
         (Math.round(b * 255) & 0xff);
}

function Update() {
  const world = getWorld();
  world.executeCommands();

  const app = Window.app;
  if (!app || !app.stage) return;
  stage = app.stage;
  stage.sortableChildren = true;

  const sortedLayers = [...layerBuckets.keys()].sort((a, b) => a - b);

  // ========== PANELS ==========
  for (const layer of sortedLayers) {
    const bucket = layerBuckets.get(layer)!;
    for (const entityIndex of bucket) {
      const gen = world.generations[entityIndex];
      const entity = new Entity(entityIndex, gen);

      if (!world.hasComponent(entity, 'Visible') || world.hasComponent(entity, 'Culled')) {
        const c = panelSprites.get(entityIndex);
        if (c) c.visible = false;
        continue;
      }

      const panel = world.getComponentByIndex<PanelComponent>(entityIndex, 'PanelComponent');
      const transform = world.getComponentByIndex<TransformComponent>(entityIndex, 'TransformComponent');
      if (!panel || !transform) continue;

      const style = styleArray[panel.styleId];
      let color = style.baseColor;
      if (world.hasComponent(entity, 'Hovered')) color = style.hoverColor ?? color;
      else if (world.hasComponent(entity, 'Selected')) color = style.selectedColor ?? color;

      let container = panelSprites.get(entityIndex);
      const w = transform.scale.x;
      const h = transform.scale.y;

      if (!container) {
        container = new PIXI.Container();
        container.label = `panel_${entityIndex}`;
        container.eventMode = 'static';
        container.on('pointerdown', (e) => {
          (window as any).__deselectImageEditor?.();
          e.stopPropagation();
        });
        stage.addChild(container);

        const bg = new PIXI.Graphics();
        bg.fill({ color: toHexColor(color.r, color.g, color.b), alpha: color.a });
        bg.rect(0, 0, w, h);
        bg.fill();
        container.addChild(bg);
        (container as any).__bg = bg;

        // clip children
        if (panel.clipChildren) {
          const mask = new PIXI.Graphics();
          mask.fill({ color: 0xffffff });
          mask.rect(0, 0, w, h);
          mask.fill();
          container.addChild(mask);
          container.mask = mask;
        }

        panelSprites.set(entityIndex, container);
        lastScaleX.set(entityIndex, w);
        lastScaleY.set(entityIndex, h);
        lastColorHex.set(entityIndex, toHexColor(color.r, color.g, color.b));
      } else {
        container.visible = true;
        const prevW = lastScaleX.get(entityIndex) ?? w;
        const prevH = lastScaleY.get(entityIndex) ?? h;
        const prevHex = lastColorHex.get(entityIndex);
        const newHex = toHexColor(color.r, color.g, color.b);
        if (prevW !== w || prevH !== h || prevHex !== newHex) {
          const bg = (container as any).__bg as PIXI.Graphics;
          if (bg) {
            bg.clear();
            bg.fill({ color: newHex, alpha: color.a });
            bg.rect(0, 0, w, h);
            bg.fill();
          }
          const existingMask = container.mask as PIXI.Graphics | undefined;
          if (existingMask) {
            existingMask.clear();
            existingMask.fill({ color: 0xffffff });
            existingMask.rect(0, 0, w, h);
            existingMask.fill();
          }
          lastScaleX.set(entityIndex, w);
          lastScaleY.set(entityIndex, h);
          lastColorHex.set(entityIndex, newHex);
        }
      }

      container.x = transform.position.x - w / 2;
      container.y = transform.position.y - h / 2;
      container.zIndex = panel.layer;
      container.alpha = panel.alpha;
    }
  }

  // ========== TEXT ==========
  world.forEachIndex2<TextComponent, TransformComponent>(
    'TextComponent',
    'TransformComponent',
    (idx, textComp, transform) => {
      const gen = world.generations[idx];
      const entity = new Entity(idx, gen);

      if (!world.hasComponent(entity, 'Visible')) {
        const txt = textSprites.get(idx);
        if (txt) txt.visible = false;
        return;
      }

      let panelVisible = true;
      if (textComp.panelIndex >= 0) {
        const pGen = world.generations[textComp.panelIndex];
        const pEntity = new Entity(textComp.panelIndex, pGen);
        panelVisible = world.hasComponent(pEntity, 'Visible');
      }
      if (!panelVisible) {
        const txt = textSprites.get(idx);
        if (txt) txt.visible = false;
        return;
      }

      let txt = textSprites.get(idx);
      if (!txt) {
        const fillColor = toHexColor(textComp.color.r, textComp.color.g, textComp.color.b);
        txt = new PIXI.Text({
          text: textComp.contentId as string,
          style: {
            fontFamily: 'Arial, sans-serif',
            fontSize: textComp.fontSize ?? 16,
            fill: fillColor,
            align: textComp.align as any,
          },
        });
        txt.label = `text_${idx}`;
        stage!.addChild(txt);
        textSprites.set(idx, txt);
      }

      txt.visible = true;
      txt.x = transform.position.x;
      txt.y = transform.position.y;
      txt.alpha = textComp.color.a;
      txt.zIndex = textComp.layer;

      const isVertical = Math.abs(textComp.rotation ?? 0) === 90 || Math.abs(textComp.rotation ?? 0) === 270;
      if (isVertical) txt.anchor.set(0.5);
      else {
        const align = textComp.align ?? 'left';
        if (align === 'center') txt.anchor.set(0.5, 0);
        else if (align === 'right') txt.anchor.set(1, 0);
        else txt.anchor.set(0, 0);
      }
      const vAlign = textComp.vAlign ?? 'top';
      if (vAlign === 'middle') txt.anchor.y = 0.5;
      else if (vAlign === 'bottom') txt.anchor.y = 1;

      txt.rotation = (textComp.rotation ?? 0) * Math.PI / 180;
    }
  );
}

export function resetDisplay() {
  for (const c of panelSprites.values()) if (c.parent) c.removeFromParent();
  for (const t of textSprites.values()) if (t.parent) t.removeFromParent();
  panelSprites.clear();
  textSprites.clear();
  lastScaleX.clear();
  lastScaleY.clear();
  lastColorHex.clear();
}

registerSystemMethod('SETUE.Scene.Scene2D', 'Update', Update);
